File: pattern-diagnose/assets/examples/typescript/adapter/before/notification-service.ts
// SPDX-License-Identifier: MIT
// Before: incompatible interface shapes — Adapter candidate

// Third-party library (cannot modify)
interface ThirdPartyLogger {
  logMessage(level: number, msg: string, context?: Record<string, unknown>): void;
}

// Our application's expected interface
interface AppLogger {
  debug(message: string): void;
  info(message: string): void;
  error(message: string, err?: Error): void;
}

// Incompatible: level is a number vs named methods, context shape differs
class NotificationService {
  // Glue code accumulates here — translation scattered at every call site
  send(message: string, thirdParty: ThirdPartyLogger): void {
    // Manual translation every time — should be centralized
    thirdParty.logMessage(1, message, { source: "notification" });
    thirdParty.logMessage(3, `notification sent: ${message}`, {
      timestamp: Date.now(),
    });
  }
}

File: pattern-diagnose/assets/examples/python/adapter/before/notification_service.py
# SPDX-License-Identifier: MIT
"""Before: incompatible interface shapes — Adapter candidate."""

from typing import Any


# Third-party library (cannot modify)
class ThirdPartyLogger:
    def log_message(self, level: int, msg: str, context: dict[str, Any] | None = None) -> None:
        print(f"[{level}] {msg} {context or ''}")


class NotificationService:
    def __init__(self, third_party: ThirdPartyLogger) -> None:
        self._logger = third_party

    def send(self, message: str) -> None:
        # Glue code accumulates here — manual translation at every call site
        self._logger.log_message(1, message, {"source": "notification"})
        self._logger.log_message(
            3, f"notification sent: {message}", {"timestamp": "now"}
        )
